File: bindings/swift/Sources/ElabifyCore/Merkle.swift
// Merkle tree construction. Byte-equivalent with @elabify/core's src/merkle.ts.
// Pads to min 8 leaves, byte-level domain tags 0x00 (empty leaf) / 0x01 (leaf) /
// 0x02 (inner). See test-vectors/merkle.kat.json + claim-leaf-hash.kat.json
// + leaf-hash.kat.json + empty-leaf-hash.kat.json for the locked contract.

import Foundation

private let tagEmptyLeaf: UInt8 = 0x00
private let tagLeaf: UInt8 = 0x01
private let tagInner: UInt8 = 0x02
private let minPaddedSize = 8

public struct MerkleProofEntry: Equatable {
    public let sibling: Data
    public let isRight: Bool
    public init(sibling: Data, isRight: Bool) {
        self.sibling = sibling
        self.isRight = isRight
    }
}

/// Spec primitive: rpo256(0x01 ‖ rpo256(keyBytes) ‖ rpo256(valueBytes)).
public func leafHash(keyBytes: Data, valueBytes: Data) -> Data {
    let kh = rpo256(keyBytes)
    let vh = rpo256(valueBytes)
    var buf = Data(capacity: 64)
    buf.append(kh)
    buf.append(vh)
    return rpo256Tagged(tagLeaf, buf)
}

/// Convenience: leaf hash from a (key, value) claim pair. Handles NFC
/// + UTF-8 encoding for the key and canonicalize() for the value.
public func claimLeafHash(key: String, value: Any) throws -> Data {
    let keyBytes = utf8Bytes(key)
    let valueBytes = try canonicalize(value)
    return leafHash(keyBytes: keyBytes, valueBytes: valueBytes)
}

/// Empty-leaf hash for padded slot index `index`: rpo256(0x00 ‖ u64BE(index)).
public func emptyLeafHash(index: UInt64) -> Data {
    var buf = Data(count: 8)
    var n = index
    for i in stride(from: 7, through: 0, by: -1) {
        buf[i] = UInt8(n & 0xff)
        n >>= 8
    }
    return rpo256Tagged(tagEmptyLeaf, buf)
}

/// Merkle inner-node combinator: rpo256(0x02 ‖ left ‖ right).
private func innerHash(_ left: Data, _ right: Data) -> Data {
    var buf = Data(capacity: left.count + right.count)
    buf.append(left)
    buf.append(right)
    return rpo256Tagged(tagInner, buf)
}

public final class MerkleTree {
    public let paddedSize: Int
    public let depth: Int
    public let root: Data
    public var rootHex: String { return bytesToHex(root) }

    private let layers: [[Data]]

    /// Build a Merkle tree over a list of (key, value) claim entries.
    /// Pads with index-tagged empty leaves up to next power of two,
    /// minimum 8.
    public init(entries: [(key: String, value: Any)]) throws {
        let leafCount = entries.count
        var target = 1
        while target < max(minPaddedSize, leafCount) { target <<= 1 }

        var leaves = [Data]()
        leaves.reserveCapacity(target)
        for entry in entries {
            leaves.append(try claimLeafHash(key: entry.key, value: entry.value))
        }
        for i in leafCount..<target {
            leaves.append(emptyLeafHash(index: UInt64(i)))
        }

        var layers: [[Data]] = [leaves]
        var cur = leaves
        while cur.count > 1 {
            var next = [Data]()
            next.reserveCapacity(cur.count / 2)
            for i in stride(from: 0, to: cur.count, by: 2) {
                next.append(innerHash(cur[i], cur[i + 1]))
            }
            layers.append(next)
            cur = next
        }
        self.layers = layers
        self.paddedSize = target
        self.depth = layers.count
        self.root = layers.last![0]
    }

    public func proof(at index: Int) -> [MerkleProofEntry] {
        precondition(index >= 0 && index < paddedSize,
                     "MerkleTree.proof: index \(index) out of range [0, \(paddedSize))")
        var entries = [MerkleProofEntry]()
        entries.reserveCapacity(layers.count - 1)
        var i = index
        for lvl in 0..<(layers.count - 1) {
            let sib = i ^ 1
            entries.append(MerkleProofEntry(
                sibling: layers[lvl][sib],
                isRight: (i & 1) == 0
            ))
            i >>= 1
        }
        return entries
    }
}

/// Verify an inclusion proof against an expected root.
public func verifyMerkleProof(
    leaf: Data,
    proof: [MerkleProofEntry],
    expectedRoot: Data
) -> Bool {
    var h = leaf
    for entry in proof {
        h = entry.isRight ? innerHash(h, entry.sibling) : innerHash(entry.sibling, h)
    }
    return h == expectedRoot
}

File: bindings/kotlin/src/main/kotlin/com/elabify/core/Merkle.kt
// Merkle tree construction. Byte-equivalent with @elabify/core's src/merkle.ts.
// Pads to min 8 leaves, byte-level domain tags 0x00 (empty leaf) / 0x01 (leaf)
// / 0x02 (inner). See test-vectors/merkle.kat.json plus the leaf-hash /
// claim-leaf-hash / empty-leaf-hash vectors for the locked contract.

package com.elabify.core

private const val TAG_EMPTY_LEAF: Int = 0x00
private const val TAG_LEAF: Int = 0x01
private const val TAG_INNER: Int = 0x02
private const val MIN_PADDED_SIZE = 8

data class MerkleProofEntry(val sibling: ByteArray, val isRight: Boolean) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is MerkleProofEntry) return false
        return isRight == other.isRight && sibling.contentEquals(other.sibling)
    }
    override fun hashCode(): Int = 31 * sibling.contentHashCode() + isRight.hashCode()
}

/** Spec primitive: rpo256(0x01 ‖ rpo256(keyBytes) ‖ rpo256(valueBytes)). */
fun leafHash(keyBytes: ByteArray, valueBytes: ByteArray): ByteArray {
    val kh = rpo256(keyBytes)
    val vh = rpo256(valueBytes)
    val buf = ByteArray(kh.size + vh.size)
    System.arraycopy(kh, 0, buf, 0, kh.size)
    System.arraycopy(vh, 0, buf, kh.size, vh.size)
    return rpo256Tagged(TAG_LEAF, buf)
}

/** Convenience: leaf hash from a (key, value) claim pair. Handles NFC +
 *  UTF-8 encoding for the key and canonicalize() for the value. */
fun claimLeafHash(key: String, value: Any?): ByteArray {
    val keyBytes = utf8Bytes(key)
    val valueBytes = canonicalize(value)
    return leafHash(keyBytes, valueBytes)
}

/** Empty-leaf hash for padded slot index [index]: rpo256(0x00 ‖ u64BE(index)). */
fun emptyLeafHash(index: Long): ByteArray {
    require(index >= 0) { "emptyLeafHash: index must be non-negative, got $index" }
    val buf = ByteArray(8)
    var n = index
    for (i in 7 downTo 0) {
        buf[i] = (n and 0xff).toByte()
        n = n ushr 8
    }
    return rpo256Tagged(TAG_EMPTY_LEAF, buf)
}

/** Merkle inner-node combinator: rpo256(0x02 ‖ left ‖ right). */
private fun innerHash(left: ByteArray, right: ByteArray): ByteArray {
    val buf = ByteArray(left.size + right.size)
    System.arraycopy(left, 0, buf, 0, left.size)
    System.arraycopy(right, 0, buf, left.size, right.size)
    return rpo256Tagged(TAG_INNER, buf)
}

class MerkleTree(entries: List<Pair<String, Any?>>) {
    val paddedSize: Int
    val depth: Int
    val root: ByteArray
    val rootHex: String get() = bytesToHex(root)

    private val layers: List<List<ByteArray>>

    init {
        val leafCount = entries.size
        var target = 1
        while (target < maxOf(MIN_PADDED_SIZE, leafCount)) target = target shl 1

        val leaves = ArrayList<ByteArray>(target)
        for (entry in entries) {
            leaves.add(claimLeafHash(entry.first, entry.second))
        }
        for (i in leafCount until target) {
            leaves.add(emptyLeafHash(i.toLong()))
        }

        val builtLayers = ArrayList<List<ByteArray>>()
        builtLayers.add(leaves.toList())
        var cur: List<ByteArray> = leaves
        while (cur.size > 1) {
            val next = ArrayList<ByteArray>(cur.size / 2)
            var i = 0
            while (i < cur.size) {
                next.add(innerHash(cur[i], cur[i + 1]))
                i += 2
            }
            builtLayers.add(next)
            cur = next
        }
        this.layers = builtLayers
        this.paddedSize = target
        this.depth = builtLayers.size
        this.root = builtLayers.last()[0]
    }

    fun proof(index: Int): List<MerkleProofEntry> {
        require(index in 0 until paddedSize) { "MerkleTree.proof: index $index out of range [0, $paddedSize)" }
        val out = ArrayList<MerkleProofEntry>(layers.size - 1)
        var i = index
        for (lvl in 0 until layers.size - 1) {
            val sib = i xor 1
            out.add(MerkleProofEntry(layers[lvl][sib], (i and 1) == 0))
            i = i shr 1
        }
        return out
    }
}

/** Verify an inclusion proof against an expected root. */
fun verifyMerkleProof(leaf: ByteArray, proof: List<MerkleProofEntry>, expectedRoot: ByteArray): Boolean {
    var h = leaf
    for (entry in proof) {
        h = if (entry.isRight) innerHash(h, entry.sibling) else innerHash(entry.sibling, h)
    }
    return h.contentEquals(expectedRoot)
}
